### ISSUE TITLE
Support rendering oneOf constraints for required properties

### ISSUE THREAD

**Issue by proppy** [mrin9/RapiDoc#254]
JSON schema support the following pattern to define valid groupings of required properties:
```
{
    "type": "object",
    "properties": {
        "emailAddress": {
            "type": "string"
        },
       "phoneNumber": {
            "type": "string"
        }
    },
    "oneOf": [
        {
            "required": [ "emailAddress" ] 
        },
        {
            "required": [ "phoneNumber" ]
        }
    ]
}
```

Currently this render in RapiDoc as:
```
...
ONEOF :
  prop1: {missing-type-info}
  prop2: {missing-type-info}
```

What do you think would be a good rendering for this? Maybe something like:
```
oneOf:
  - required: *emailAddress
  - required: *phoneNumber
```

**Comment by mrin9** [comment]
Thanks for bringing it up,
Will be holding discussion on it. This is how I imagine it to be shown

**Schema**
```
ONE-OF:
1: {
     emailAddress*: string
   }
2: {
     phoneNumber*: string
   }
```
**Example**
```
{
  emailAddress: "string"
}
```

I am open to take a PR on this

**Comment by proppy** [comment]
So we would effectively merge the `oneOf` `required` constraints w/ the outer object as if the `properties` definitions was in the `oneOf`?

Turning this:
```
{
    "type": "object",
    "properties": {
        "emailAddress": {
            "type": "string"
        },
       "phoneNumber": {
            "type": "string"
        }
    },
    "oneOf": [
        {
            "required": [ "emailAddress" ] 
        },
        {
            "required": [ "phoneNumber" ]
        }
    ]
}
```
into something equivalent to:
```
{

    "oneOf": [
        {
            "type": "object",
            "properties": {
                "emailAddress": {
                    "type": "string"
                }
            },
            "required": [ "emailAddress" ] 
        },
        {
            "type": "object",
            "properties": {
               "phoneNumber": {
                    "type": "string"
                }
            },
            "required": [ "phoneNumber" ]
        }
    ]
}
```

**Comment by mrin9** [comment]
Some question raised during our discussion 

with no `type/property/item` information is it a valid JSON schema ?  
```
{
    "required": [ "phoneNumber" ]
 }
```
per the [JSON schema spec ](https://json-schema.org/draft/2019-09/json-schema-core.html#rfc.section.9.2.1.3)
each item of `oneof` array must be a valid json schema.

can you point me to a location where it would confirm if just having a `required` property makes the object a valid JSON schema ?

**Comment by proppy** [comment]
Looking at the meta-schema for validation: http://json-schema.org/draft/2019-09/schema it doesn't look like any of they keyword is actually required.

**Comment by mrin9** [comment]
>My understanding is that a schema like:
{
  "$schema": "https://json-schema.org/draft/2019-09/schema",
  "required": ["foo"]
}
would simply means that a "foo" property is required (independently of any type constraints on the actual property).

I think thats where RapiDoc is diffing in its interpretation, 
```
{
  "required": ["foo"]
}
```

the way RapiDoc would like to think about the above definition is,  __items in `required` array is valid only when they are defined on the instance.__ 
so the above schema def per RapiDoc would yield to 
```
{ 
   
}
```
If items in required array aren't defined we wont have any `type` for `foo` and that would mean that it can be anything 
a primitive, null, object or array and representing something without a type will be very dogmatic.

so if I am not wrong your interpretation of it should look like 
```
{
  foo : <missing type info>
}
``` 

>PS: This issue is still under discussion, and we are open to changes/suggestions

**Comment by proppy** [comment]
> so if I am not wrong your interpretation of it would yield to,
>  ```
> {
>   foo : <missing type info>
> }
> ```

Maybe it should be:
```
{
   foo*: <missing type info>
}
```
to capture that is is required (notice the `*` before the colon).

But then it's unclear what rapidoc should do when the properties type are defined earlier in the root schema. Showing duplicate information may create more confusion and feel a bit redundant:
```
{
  emailAddress: string
  phoneNumber: string
}
ONE-OF:
1: {
     emailAddress*: string
   }
2: {
     phoneNumber*: string
   }
```
> *items in required array is valid only when they are defined on the instance.*

I think this is aligned with the JSON schema spec wording (or at least my interpretation of it).

Maybe rapidoc could have a way to render this particular contraint: `valid only when they are defined on the instance.` separately from (or without knowning about) its type.
```
{
  emailAddress: string
  phoneNumber: string
}
ONE-OF:
1: {
      emailAddress*
   }
2: {
      phoneNumber*
   }
```

**Comment by bkarvas** [comment]
Highly looking for support of this feature.  
ReDoc supports both definition options.

> ```
> {
>     "type": "object",
>     "properties": {
>         "emailAddress": {
>             "type": "string"
>         },
>        "phoneNumber": {
>             "type": "string"
>         }
>     },
>     "oneOf": [
>         {
>             "required": [ "emailAddress" ] 
>         },
>         {
>             "required": [ "phoneNumber" ]
>         }
>     ]
> }
> ```
> 
> ```
> {
> 
>     "oneOf": [
>         {
>             "type": "object",
>             "properties": {
>                 "emailAddress": {
>                     "type": "string"
>                 }
>             },
>             "required": [ "emailAddress" ] 
>         },
>         {
>             "type": "object",
>             "properties": {
>                "phoneNumber": {
>                     "type": "string"
>                 }
>             },
>             "required": [ "phoneNumber" ]
>         }
>     ]
> }
> ```

**Comment by mrin9** [comment]
@bkarvas, we do support the 2nd, but I think the discussion here is on how to interpret the first schema def.
does Redoc produces the same example output with both definition ?

For the 1st schema swagger produces 
```
{
  emailAddress: string
  phoneNumber: string
}
ONE-OF:
1: {
   }
2: {
   }
```

and the an example for the schema would be 
```
{
  emailAddress: "string",
  phoneNumber: string
}
```
which majority of our team thinks is in more inline with json schema wording. we are inclining to go this route
if you have any opinion on it, then now is the time to bring it up.

> CC: @proppy

**Comment by bkarvas** [comment]
@mrin9, yes, ReDoc produces the same result for both definition. Look at the pictures below. From my point of view you can interpret the 1st schema the same way as you do with the 2nd

![Sample1](https://user-images.githubusercontent.com/66029993/87297407-ab259a80-c508-11ea-99bf-2bc0e1347350.png)

![Sample2](https://user-images.githubusercontent.com/66029993/87297431-b1b41200-c508-11ea-86b4-200655cccdd1.png)

Sample YAML:
```
openapi: '3.0.2'
info:
  title: RapiDoc Issue 253 & 259
  version: '1.0'
paths:
  /test1:
    post:
      requestBody:
        content:
          application/json:
            schema:
              $ref: '#/components/schemas/Amount1'
        description: JSON request body for new payment initiation API.
        required: true

      responses:
        '200':
          description: OK

        "400":
          description: Bad request.
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Errors'
              example:
                errors: 
                - error: 'INVALID_REQUEST'
                  message: 'Invalid request.'              
        "401":
          description: Unauthorized.
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Errors'
              example: |
                {
                  "errors": [
                  {
                    "error": "UNAUTHORIZED",
                    "message": "Attempt to call a method without a proper licence or invalid certificate."
                  }
                  ]
                }

  /test2:
    post:
      requestBody:
        content:
          application/json:
            schema:
              $ref: '#/components/schemas/Amount2'
        description: JSON request body for new payment initiation API.
        required: true

      responses:
        '200':
          description: OK
        "400":
          description: Bad request.
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Errors'
              example:
                errors: 
                - error: 'INVALID_REQUEST'
                  message: 'Invalid request.'              
        "403":
          description: Forbidden.
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Errors'
              examples:
                example-1:
                  value:
                    errors:
                    - error: FORBIDDEN 1
                      message: No disposal rights.
                example-2:
                  value:
                    errors:
                    - error: 'FORBIDDEN 2'
                      message: 'No disposal rights.'

components:
  schemas:
    Amount1:
      description: Payment amount.
      example:
        equivalentAmount:
          currency: USD
          value: 1245.50
        instructedAmount:
          currency: USD
          value: 1245.50
      oneOf:
        - properties:          
            instructedAmount:
                $ref: '#/components/schemas/InstructedAmount'
            equivalentAmount:
              $ref: '#/components/schemas/EquivalentAmount'
          required:
            - instructedAmount         
        - properties: 
            instructedAmount:
                $ref: '#/components/schemas/InstructedAmount'
            equivalentAmount:
              $ref: '#/components/schemas/EquivalentAmount'
          required:
            - equivalentAmount         

    Amount2:
      description: Payment amount.
      example:
        equivalentAmount:
          currency: USD
          value: 1245.50
        instructedAmount:
          currency: USD
          value: 1245.50
      oneOf:
      - required:
        - instructedAmount
      - required:
        - equivalentAmount
      properties:
        instructedAmount:
          $ref: '#/components/schemas/InstructedAmount'
        equivalentAmount:
          $ref: '#/components/schemas/EquivalentAmount'
      type: object

    InstructedAmount:
      description: Amount and currency in the instruction.
      example:
        currency: USD
        value: 1245.50
      properties:
        value:
          description: The amount value.
          example: 1245.50
          maximum: 9999999999999.99
          minimum: 0
          multipleOf: 0.01
          type: number
        currency:
          description: Amount currency. Currency code according to ISO 4217 = 3 capital
            letters.
          example: USD
          maxLength: 3
          pattern: ^[A-Z]{3}$
          type: string
      required:
      - currency
      - value
      type: object
    EquivalentAmount:
      description: Equivalent amount and currency.
      example:
        currency: USD
        value: 1245.50
      properties:
        value:
          description: The amount value.
          example: 1245.50
          maximum: 9999999999999.99
          minimum: 0
          multipleOf: 0.01
          type: number
        currency:
          description: Amount currency. Currency code according to ISO 4217 = 3 capital
            letters.
          example: USD
          maxLength: 3
          pattern: ^[A-Z]{3}$
          type: string
      required:
      - currency
      - value
      type: object

    Errors:
      type: object
      description: Collection of application errors.
      properties:
        errors:
          description: List of application errors.
          type: array
          items:
            $ref: '#/components/schemas/Error'
    Error:
      description: An application error item.
      type: object
      required:
      - error
      properties:
        error:
          description: Error code.
          example: ERROR_CODE
          type: string
        parameters:
          description: Error parameters.
          type: object
        scope:
          description: Error scope.
          type: string
        message:
          description: Errro detail.
          type: string
          example: Invalid request.

```

**Comment by mrin9** [comment]
thanks @bkarvas for your input and the screenshots. I will keep this open for some time for other to leave their feedback/opinion about it. We will also discuss within the team and then provide a fix

**Comment by bkarvas** [comment]
Any progress on this issue?

**Comment by mrin9** [comment]
We discussed this issues within our team and came to a consensus that the way the incomplete schema is represented in `<rapi-doc> v8.3.0` is good enough.

To expand a bit more on that. We are talking about how to represent the following schema in the spec 

```yml
openapi: 3.0.0
info:
  version: 0.0.1
  title: Testing incomplete schema 
paths:
  /some-path:
    get:
      summary: to test
      responses:
        '200':
          description: Successful response
            application/json:
              schema:
                type: object
                properties:
                  emailAddress:
                    type: string
                  phoneNumber:
                    type: string
                oneOf:
                  - required:
                      - emailAddress
                  - required:
                      - phoneNumber
```

Apart from discussing within our team that, we have checked with many other commercial and open-sourced  openapi renderers. some didnt work and some representation did not make much sense.

One of the representation of the above schema was from Swagger-UI which aligns with our interpretation of this.
for an comparison below images shows how its represented in Swagger UI and RapiDoc

Swagger
<kbd>
<img width="599" alt="Screen Shot 2020-11-12 at 9 40 31 AM" src="https://user-images.githubusercontent.com/5056899/98977923-07f08c00-24ce-11eb-8df8-c467a0609133.png">
</kbd>

RapiDoc
<kbd>
<img width="541" alt="Screen Shot 2020-11-12 at 9 39 32 AM" src="https://user-images.githubusercontent.com/5056899/98977932-0b841300-24ce-11eb-91f4-364d9cf60bb9.png">
</kbd>

I hope it works for everyones use-case
